### PR TITLE
OLMIS-3049 OLMIS-3095 Allow filtering by multiple values in requisitionsForConvert

### PR DESCRIPTION
**Pull request by sebbrudzinski**
The endpoint now allows providing multiple values in GET param
"filterValue". If multiple values are provided the endpoint will
look for any match for all the provided values. If no value is provided
the endpoint still doesn't perform any filtering.

---
This change is [Reviewable](https://reviewable.io/reviews/openlmis/openlmis-requisition/44)

### CODE REVIEW

**Comment by sebbrudzinski** [comment]
Hey guys. I didn't manage to get any feedback on this contribution, so I'm submitting it as a pull request. Please review.

File: src/main/java/org/openlmis/requisition/service/RequisitionService.java [hunk 1]
```diff
@@ -647,32 +648,66 @@ private boolean approvedQtyColumnEnabled(Requisition requisition) {
...
+
+    facilities.forEach(facilityDto -> uuidsToReturn.add(facilityDto.getId()));
+    programs.forEach(programDto -> uuidsToReturn.add(programDto.getId()));
+
+    return uuidsToReturn;
+  }
+
+  private Collection<ProgramDto> findProgramsWithFilter(String filterBy,
+                                                        List<String> filterValues) {
+    if (filterValues.isEmpty()) {
```
Comment (pgesek) [inline src/main/java/org/openlmis/requisition/service/RequisitionService.java]: shouldn't we do this after the check on line 673?

File: src/test/java/org/openlmis/requisition/service/RequisitionServiceTest.java [hunk 2]
```diff
@@ -1037,6 +1037,43 @@ public void shouldFindAndSortAndPageApprovedRequisitions() {
   }
 
   @Test
+  public void shouldFilterRequisitionsForConvertByMultipleValues() {
```
Comment (pgesek) [inline src/test/java/org/openlmis/requisition/service/RequisitionServiceTest.java]: we should add a test for facility search as well. "all' would be good to test as the param as well

File: src/main/java/org/openlmis/requisition/service/RequisitionService.java (hunk 1 above)
Comment (sebbrudzinski) [inline src/main/java/org/openlmis/requisition/service/RequisitionService.java]: Yes, definitely. Good catch.

File: src/test/java/org/openlmis/requisition/service/RequisitionServiceTest.java (hunk 2 above)
Comment (sebbrudzinski) [inline src/test/java/org/openlmis/requisition/service/RequisitionServiceTest.java]: Done

File: src/main/java/org/openlmis/requisition/service/RequisitionService.java (hunk 1 above)
Comment (sebbrudzinski) [inline src/main/java/org/openlmis/requisition/service/RequisitionService.java]: Done. Also there was a problem with how it worked for facilities.

**Comment by joshzamor** [comment]
Hi Sebastian, I've only had time to give this the most basic of looks.  My overall question/concern is that this looks like the filtering is done in Java code.  Is that right?  If so what's the maximum size of the collection that this java code would need to filter?

**Comment by sebbrudzinski** [comment]
@joshzamor That's partially correct. What's done on the Java code side is retrieving the UUIDs of programs/facilities that match specific criteria (like program name or facility name/code). Once we have those UUIDs, they are passed to the database query to only retrieve those approved requisitions that match the facility/program requested.

**Comment by joshzamor** [comment]
@sebbrudzinski :  thanks.  I'd much prefer less filtering in Java code.  We've seen significant (10x) performance improvements by not filtering in Java code when applied at scale.  Given this is an operation that I've heard has caused performance problems, I'm concerned how we don't know how this would effect this endpoint.  When I run the original requisitionsForConvert on the performance testing data (60 second timeout and far less data than Malawi), it times out.

Could you do some manual testing, with Malawi's data, e.g. on Malawi's data, and give us a sense of what the performance impact is?

**Comment by sebbrudzinski** [comment]
@joshzamor Yes, we have performance problems with this endpoint and this change is actually meant to improve our convert to order screen. By default it fetches and processes all requisitions in 'approved' status. The number of legacy approved requisitions for non-ordering programs is about 150k and this endpoint has to process them all if used as is. By implementing this ability to filter by multiple values, we were planning to use this endpoint to only give us results for ordering programs and therefore reducing the number of requisitions it needs to process to about 500 at max. 

I honestly don't see how we can have even less filtering in java code than we do currently if we want to keep the current filtering behaviour. All we do is contact the referencedata service to give us matching program and facility UUIDs to use them in query later on. One way to do it would be to completely remove filtering by user specified strings and make the endpoint accept UUID of the program/facility. This however would require a re-work of the search box that currently exists in core UI. Is this second option something you would rather see? (refactoring the endpoint not to allow custom-string searches, but just allow to provide facility/program uuids). 

I can certainly do some manual testing with the new approach. I can for sure say that when it was processing 150k requisitions (no filtering) it didn't finish within 15minutes time frame.

**Comment by sebbrudzinski** [comment]
@joshzamor I ran some tests on the Malawi data and wrote down the results in the ticket https://openlmis.atlassian.net/browse/MW-534

**Comment by pgesek** [comment]
Looking at the code the changes shouldn't have drastic performance impact, given that we only loop through the number of parameters, which is reflected in the results.

A further improvement would be making ref data accept multiple filter values, which would remove the need to loop, but that's peanuts.

The biggest performance sink that is obvious to me in this endpoint is the in-memory pagination and the getAvailableSupplyDepots call that is being made for each of the retrieved requisitions. Any other changes won't have meaningful impact as long as that's there.

**Comment by pgesek** [comment]
A one liner bonus performance gain by the way is switching the code in searchApprovedRequisitionsWithSortAndFilterAndPaging to create the dto list **after** doing the filtering, not before.

**Comment by sebbrudzinski** [comment]
@pgesek at your latter comment - I think it is done after filtering - we only create DTOs for requisitions that the query returned for our matching UUIDs. This is done before sorting and paginating though (due to the need to sort by properties that come from ref data)

**Comment by sebbrudzinski** [comment]
I think adjusting calls to retrieve the availableSupplyingDepots is within the reach and I'll be happy to do that.

**Comment by pgesek** [comment]
@sebbrudzinski it's done after the original filtering, but before filtering by supplying depots.

**Comment by sebbrudzinski** [comment]
@pgesek https://openlmis.atlassian.net/browse/OLMIS-3095 I will work on that tomorrow.

**Comment by sebbrudzinski** [comment]
Are there any more comments related to this change?

**Comment by sebbrudzinski** [comment]
Added performance improvements mentioned above @pgesek 
If this won't be making it into the release, I'll need an info by tomorrow morning.

**Comment by joshzamor** [comment]
@sebbrudzinski,thanks for the work here, and the performance analysis.  I think you're right, this will have a net positive impact.  Lets merge it.

File: src/integration-test/java/org/openlmis/requisition/web/RequisitionControllerIntegrationTest.java [hunk 3]
```diff
@@ -1163,11 +1168,59 @@ public void shouldGetApprovedRequisitionsWithSortByFilterByAndPaging() {
...
+        .statusCode(200)
+        .extract().as(PageImplRepresentation.class);
+
+    // then
+    assertEquals(1, response.getContent().size());
+    assertThat(RAML_ASSERT_MESSAGE, restAssured.getLastReport(), RamlMatchers.hasNoViolations());
+  }
+
+  @Test
+  public void shouldGetApprovedRequisitionsWithSortByMultipleFilterValues() {
```
Comment (palbecki) [inline src/integration-test/java/org/openlmis/requisition/web/RequisitionControllerIntegrationTest.java]: Can you remove code duplication from this test and above?

Comment (sebbrudzinski) [reply src/integration-test/java/org/openlmis/requisition/web/RequisitionControllerIntegrationTest.java]: Which parts do you think would make sense extracting as private methods, such that they do not interrupt usage of variables in the test? @palbecki

Comment (palbecki) [reply src/integration-test/java/org/openlmis/requisition/web/RequisitionControllerIntegrationTest.java]: Can we extract everything and pass to method param filterValue? If not then just extract lines that are duplicated for few private methods

Comment (sebbrudzinski) [reply src/integration-test/java/org/openlmis/requisition/web/RequisitionControllerIntegrationTest.java]: I extracted right mocking and reused facility mocking. Anything else?

Comment (palbecki) [reply src/integration-test/java/org/openlmis/requisition/web/RequisitionControllerIntegrationTest.java]: looks good, thanks

**Comment by sebbrudzinski** [comment]
@palbecki @mkwiatkowskisoldevelo @pgesek Anything else before we merge it in?

**Pull request merged**